python/ray/workflow/http_event_provider.py

```diff
@@ -0,0 +1,240 @@
+from ray.workflow.common import Event
+
+class WorkflowEventHandleError(Exception):
+    def __init__(self, workflow_id: str, what_happen: str):
+        self.message = f"Workflow[id={workflow_id}] HTTP event handle failed: {what_happen}"
+        super().__init__(self.message)
+
+app = FastAPI()
+
+@serve.deployment(name="myEventProvider", route_prefix="/event")
+@serve.ingress(app)
+class HTTPEventProvider():
+    """HTTPEventProvider is defined to be a Ray Serve deployment with route_prefix='/event',
+    which will receive external events via an HTTP endpoint. It supports FastAPI, e.g. post.
+    It responds to both poll_for_event() and event_checkpointed() from an HTTPListener instance
+    or an HTTPListenerInEventCoordinator instance.
+
+    NOTE:
+    HTTPListener class is designed to work with the current workflow.wait_for_event()
+    implementation, where an HTTPListener instance will be initiated by the get_message() and
+    message_committed() of the workflow.wait_for_event().
+
+    HTTPListenerInEventCoordinator, on the other hand, is intended to work withh a
+    WorkflowEventCoordinator and a revised workflow.wait_for_event() implementation.
+
+    Example definition
+    ==================
+
+    ```
+    class HTTPEventProvider:
+
+        def __init__(self, event_coordinator_ack: Optional[bool] = False):
+
+        @app.post("/arrivals")
+        async def create_event_payload(self, req: Request):
+
+        async def get_event_payload(self, workflow_id, event_key):
+
+        async def report_checkpointed(self, workflow_id, event_key, confirmation):
+
+    ```
+
+    Example Usage
+    =============
+    >>> from ray.workflow.http_event_provider import HTTPEventProvider, HTTPListener
+    >>> ray.init(address='auto', namespace='serve')
+    >>> serve.start(detached=True)
+    >>> HTTPEventProvider.deploy()
+    >>> event_step = workflow.wait_for_event( # doctest: +SKIP
+    ...     HTTPListener, workflow_id='', event_key='')
+    >>> handle_event = ... # doctest: +SKIP
+    >>> handle_event.step(event_step).run() # doctest: +SKIP
+    >>>
+    >>> On a separate python process, it sends an event to the HTTPEventProvider.
+    >>> import requests
+    >>> resp = requests.post('http://127.0.0.1:8000/event/arrivals',
+    ...     json={'workflow_id':'my_wf','event_key':'my_key','event_payload':'testMessage'})
+    >>>
+
+    """
+
+    def __init__(self, event_coordinator_ack: Optional[bool] = False):
+        import nest_asyncio
+        nest_asyncio.apply()
+        self.event_key_payload: Dict[str, Dict[Any, Any]] = {}
+        self.event_coordinator_ack = event_coordinator_ack
+        if self.event_coordinator_ack:
+            self.event_key_checkpointed: Dict[str, Dict[Any, Any]] = {}
+        else:
+            self.payload_record: Dict[Any] = {}
+
+    @app.post("/arrivals")
+    async def create_event_payload(self, req: Request):
+        req_json = await req.json()
```

Comment:
Why do we need await for json? https://requests.readthedocs.io/en/latest/user/quickstart/#json-response-content

Reply:
Fastapi uses starlette underneath. This Request object is from starlette. 
https://www.starlette.io/requests/
Starlette APIs are async methods. 
The request body as bytes: await request.body()
The request body, parsed as form data or multipart: await request.form()
The request body, parsed as JSON: await request.json()

Reply:
Got it.